src/main/java/com/aws/greengrass/mqtt/bridge/clients/IoTCoreClient.java

```diff
@@ -9,76 +9,66 @@
-            RetryUtils.RetryConfig.builder().initialRetryInterval(Duration.ofSeconds(1L))
-                    .maxRetryInterval(Duration.ofSeconds(120L)).maxAttempt(Integer.MAX_VALUE)
-                    .retryableExceptions(Arrays.asList(ExecutionException.class, TimeoutException.class)).build();
-
-    @Getter(AccessLevel.PROTECTED)
-    private Set<String> subscribedIotCoreTopics = ConcurrentHashMap.newKeySet();
-    @Getter(AccessLevel.PROTECTED)
-    private Set<String> toSubscribeIotCoreTopics = new HashSet<>();
+    private static final String LOG_KEY_TOPIC = "topic";
+    private static final String LOG_KEY_TOPICS = "topics";
+
+    private final SubscriptionManager subscriptionManager = new SubscriptionManager();
     private volatile Consumer<com.aws.greengrass.mqtt.bridge.model.MqttMessage> messageHandler;
-    private Future<?> subscribeFuture;
-    private final Object subscribeLock = new Object();
+
     private final MqttClient iotMqttClient;
     private final ExecutorService executorService;
 
-    private final Consumer<MqttMessage> iotCoreCallback = (message) -> {
+    private final Consumer<Publish> iotCoreCallback = (message) -> {
         String topic = message.getTopic();
-        LOGGER.atTrace().kv(TOPIC, topic).log("Received IoT Core message");
+        LOGGER.atTrace().kv(LOG_KEY_TOPIC, topic).log("Received IoT Core message");
 
         Consumer<com.aws.greengrass.mqtt.bridge.model.MqttMessage> handler = messageHandler;
         if (handler == null) {
-            LOGGER.atWarn().kv(TOPIC, topic).log("IoT Core message received but message handler not set");
+            LOGGER.atWarn().kv(LOG_KEY_TOPIC, topic).log("IoT Core message received but message handler not set");
         } else {
-            handler.accept(com.aws.greengrass.mqtt.bridge.model.MqttMessage.fromCrtMQTT3(message));
+            handler.accept(com.aws.greengrass.mqtt.bridge.model.MqttMessage.fromSpoolerV5Model(message));
         }
     };
 
     @Getter(AccessLevel.PACKAGE) // for unit testing
     private final MqttClientConnectionEvents connectionCallbacks = new MqttClientConnectionEvents() {
         @Override
         public void onConnectionInterrupted(int errorCode) {
+            subscriptionManager.offline();
         }
 
         @Override
         public void onConnectionResumed(boolean sessionPresent) {
-            synchronized (subscribeLock)  {
-                if (subscribeFuture != null) {
-                    subscribeFuture.cancel(true);
-                }
-                // subscribe to any topics left to be subscribed
-                Set<String> topicsToSubscribe = new HashSet<>(toSubscribeIotCoreTopics);
-                topicsToSubscribe.removeAll(subscribedIotCoreTopics);
-                subscribeFuture = executorService.submit(() -> subscribeToTopicsWithRetry(topicsToSubscribe));
-            }
+            subscriptionManager.online();
+            subscriptionManager.subscribeAll();
```

Comment:
note that if we've already subscribed successfully the spooler will handle the resubscription (if needed). If we haven't subscribed successfully yet, then yes you need to retry.